### PR TITLE
Action Form Validation

### PR DESCRIPTION
**Pull request by iethree**
resolves [#26579](https://github.com/metabase/metabase/issues/26579)

## Description
- Uses new `optional` prop in action form components from https://github.com/metabase/metabase/pull/27246
- auto-generates `Yup` form validation schema from form settings, and enforces form validation
- removes bespoke `setDefaultValues` and `setNumericValues` functions from form submission in favor of the `Yup` implementations

Screen Shot 2022-12-15 at 2 46 41 PM

liveupdate

## Testing steps: implicit action forms
- Within a data app, add a scaffolded page
- click on one of the items to go to the detail page
- open up the edit form for that item
- the form should correctly identify which fields are optional (from the db)
- the form should not be submittable with any required fields empty
- the form should be submittable with required fields filled

## Testing steps: custom action forms
- Open a custom action with a query with one or more parameters
- see that toggling the required field setting in fact toggles the (optional) label int he preview
- add the action to a dashboard, and see that the optional labels appear appropriately
- see that the form can only be submitted when all optional fields are provided
- see that linking a dashboard filter to one or more fields in the form still allows the form to be submitted.

### CODE REVIEW

**Review by alxnddr — COMMENTED** [review]
@iethree looks good! I've got a couple of questions while looking at the PR. Maybe some of them are not directly related to this exact PR, but it would really help me to understand data apps better.

1) Non-null booleans marked as optional — is this intentional?
2) Why do we require optional fields when submitting a custom form?
3) For nullable values, how do I set null?
4) An empty string is a valid value of non-nullable text column but I can't set it for a required field. Is this intentional?
5) How do I remove a value of an optional datetime field? When I clear the input, submitting a form fails

**Comment by iethree** [comment]
> @iethree looks good! I've got a couple of questions while looking at the PR. Maybe some of them are not directly related to this exact PR, but it would really help me to understand data apps better.

> 1. Non-null booleans marked as optional — is this intentional?
the "(optional)" tag reflects whether the database requires a value (is the field nullable), so I think this is technically correct, but you're right that our boolean form component displays "false" and "null" the same:

Screen Shot 2022-12-21 at 2 38 16 PM
Screen Shot 2022-12-21 at 2 38 09 PM

I'm not exactly sure we currently have a better way to handle this?

> 2. Why do we require optional fields when submitting a custom form?

I'm not 100% sure what you mean here, In my own testing, the optional fields do not need to be provided. But my best guess is that if you ran into this error form the backend that requires at least one changed field? I think I can handle this better, and block form submission if no parameters have been changed.

Screen Shot 2022-12-21 at 2 40 04 PM

> 3. For nullable values, how do I set null?

I have no idea. 🙃  I think we should get some product guidance here.

> 4. An empty string is a valid value of non-nullable text column but I can't set it for a required field. Is this intentional?

With our old forms framework it automatically cast all `null` to `''`, I think we can do better now actually.

> 5. How do I remove a value of an optional datetime field? When I clear the input, submitting a form fails

Let me look into this one, I think `Yup` might be able to help us here too.

**Comment by iethree** [comment]
@alxnddr Thanks for the great feedback. Here's where this landed
1. Until we have a UI and/or product guidance for setting null vs empty strings, all nullable fields will cast to null when empty. Without a different UI component, we can't set nullable booleans to null for now. I'll throw this on product's radar, but i suspect this is too deep in the weeds to sort out right now.
2. Action forms cannot be submitted unless at least one field has been changed now.

This behavior is now concretized in unit tests, so if/when we update it can be clear how the behavior has changed.

File: frontend/src/metabase/actions/components/ActionForm/ActionForm.unit.spec.tsx [hunk 1]
```diff
@@ -300,6 +298,398 @@ describe("Actions > ActionForm", () => {
...
+          {
+            "abc-123": null,
+          },
+          expect.any(Object),
+        );
+      });
+    });
+
+    // bug repro: https://github.com/metabase/metabase/issues/27377
+    it.skip("casts empty optional category fields to null", async () => {
```
Comment (iethree) [inline frontend/src/metabase/actions/components/ActionForm/ActionForm.unit.spec.tsx:654]: @nemanjaglumac I added my own bug repro 😂

**Comment by alxnddr** [comment]
This is what I meant by requiring optional fields:
<video src="https://user-images.githubusercontent.com/14301985/209194828-812555ac-9df6-4024-81af-6681725e7a8c.mov" />

**Comment by alxnddr** [comment]
@iethree is this because the field technically does not have a default value? If we do not allow submitting such forms that do not have default values, should we restrict saving them like that?

File: frontend/src/metabase/actions/components/ActionForm/utils.ts [hunk 2]
```diff
@@ -93,3 +94,50 @@ export const getForm = (
...
+const getFieldValidationType = (fieldSettings: FieldSettings) => {
+  switch (fieldSettings.inputType) {
+    case "number":
+      return Yup.number();
+    case "boolean":
+      return Yup.boolean();
+    case "date":
+    case "datetime":
+    case "time":
+      // for dates, cast empty strings to null
```
Comment (alxnddr) [inline frontend/src/metabase/actions/components/ActionForm/utils.ts:107]: 👍

**Review by alxnddr — APPROVED** [review]
Looks great, thanks for addressing my feedback

I created a couple of other issues based on our discussion and my findings:
https://github.com/metabase/metabase/issues/27384
https://github.com/metabase/metabase/issues/27386

Please also create more issues for missing things you find

**Comment by iethree** [comment]
> This is what I meant by requiring optional fields:
> 
>  require.optional.fields.mov

oh yes, you ran into https://github.com/metabase/metabase/issues/26588, which requires [[ square brackets ]] around optional parameters which is horrendous UX and we need to find a better way to handle.

**Pull request merged**